src/FluentValidation/Validators/RangeValidator.cs

```diff
@@ -0,0 +1,63 @@
+﻿#region License
+// Copyright (c) .NET Foundation and contributors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// The latest version of this file can be found at https://github.com/FluentValidation/FluentValidation
+#endregion
+
+namespace FluentValidation.Validators {
+	using System;
+	using System.Collections.Generic;
+
+	public abstract class RangeValidator<T, TProperty> : PropertyValidator<T, TProperty>, IBetweenValidator {
+
+		public RangeValidator(TProperty from, TProperty to) {
+			To = to;
+			From = from;
+
+			if (Compare(to, from) == -1) {
+				throw new ArgumentOutOfRangeException(nameof(to), "To should be larger than from.");
+			}
+		}
+		public RangeValidator(TProperty from, TProperty to, IComparer<TProperty> comparer) {
+			To = to;
+			From = from;
+
+			explicitComparer = comparer;
+
+			if (comparer.Compare(to, from) == -1) {
+				throw new ArgumentOutOfRangeException(nameof(to), "To should be larger than from.");
+			}
+		}
+
+		readonly IComparer<TProperty> explicitComparer = null;
+
+		public TProperty From { get; }
+		public TProperty To { get; }
+
+		object IBetweenValidator.From => From;
+		object IBetweenValidator.To => To;
+
+		protected int Compare(TProperty a, TProperty b) {
+
+			// Use explicitComparer first
+			if (explicitComparer != null)
+				return explicitComparer.Compare(a, b);
+			else if (a is IComparable comparableA)

```

Comment:
Using the non-generic `IComparable` here is going to cause unnecessary boxing when working with value types - it would be better if we could find a way to use `IComparable<TProperty>` instead, as per the previous implementation.

Reply:
It could be that we have to have 2 versions of the validators - one for handling `IComparable<T>` and one with the comparer.

Reply:
The other option would be to always work with `IComparer<T>` and have a comparer implementation that explicitly wraps comparing two `IComparable<T>` instances.

Reply:
Eg: 
```csharp
	public class ComparableComparer<T> : IComparer<T> where T : IComparable<T> {
		public int Compare(T x, T y) {
			return x.CompareTo(y);
		}
	}

```

This could then be passed in via the extension methods that explicitly work with IComparable.

Reply:
Went through a few iterations and I think this solution is not too bad. There are no breaking changes.

Both `ComparableComparer` and 2 versions had to be implemented to accommodate direct instantiation ```new InclusiveBetweenValidator<Person, int>(1, 10);``` calls.

Because of 4 different validators, I moved `IsValid` implementation down to `RangeValidator` and introduced HasError for the check. Not so sure about it as it could be confusing.

Some other thoughts/considerations:
- In my project there is a need to store degrees. The valid range is [0, 360) - meaning 0 is inclusive, but 360 isn't. If that's a going to be a popular request in the future, that means 2 more validators.
- Perhaps RangeValidator could take an Enum Inclusive, Exclusive, InclusiveFromExclusiveTo (for lack of a better name)
- Should be a separate issue probably: The "Value" argument is different from "PropertyValue" in the [reference](https://docs.fluentvalidation.net/en/latest/built-in-validators.html#inclusivebetween-validator)

Reply:
Thanks, I'll have another review of this tomorrow. I might end up making a few tweaks and I'll let you know if I do.

>  If that's a going to be a popular request in the future, that means 2 more validators.

To be honest I probably wouldn't bother adding new classes for this, as it's easy enough to use `Must` to perform the comparison in this case.  

> "Value" argument is different from "PropertyValue" in the reference

These are different - {PropertyValue} is automatically added to all validators. The {Value} placeholder is the comparison value (and should actually be called {ComparisonValue} for consistency with other validators). I'll open a separate issue to rename this and add it to the docs.

Reply:
Actually I was wrong about ComparisonValue - the {Value} placeholder in these 2 validators is a duplication of {PropertyValue}, so it should be possible to remove it, but I may leave it in place for backwards compatibility